### ISSUE TITLE
1.0.7 breaks routing in Laravel 5.1.20

### ISSUE THREAD

**Issue by martin-georgiev** [laravel-doctrine/orm#80]
A change in 1.0.7 introduced unintentional bug in Laravel's routing mechanism.
If a route is defined using the class keyword for resolving the class name like here
`Route::resource('some-route', \App\Http\Controllers\Controller::class);`
Laravel will try to resolve a file from namespace `App\Http\Controllers\Controller\App\Http\Controllers\Controller` instead of `\App\Http\Controllers\Controller`.
On the other hand passing a string containing the fully qualified namespace of the file will work and will not break the routing.

**Comment by maxbrokman** [comment]
It seems unlikely that this is an issue with LD.

Are you calling `Route::resource()` from within a route group, and if so does it have a namespace set?

E.g by default in 5.1 your `RouteServiceProvider` comes with

``` php
$router->group(['namespace' => $this->namespace], function ($router) {
    //
}
```

In that case using `Controller::class` won't work because Laravel will append a namespace string to the front of whatever you pass in.

**Comment by martin-georgiev** [comment]
With `1.0.6` the issue cannot be reproduced. It appears only with `1.0.7`.
I reckon some change in `1.0.7` triggers different workflow in Laravel and this is the final result.
It is a confirmed issue with Laravel 5.1.20 and 5.1.22.

The issue is observable with other routes definitions and not only when `Route::resource()` is used.
The side effect is presented for sure with `::controller`, `::resource`, `::get`, `::post`.

**Comment by guiwoda** [comment]
Can you provide a complete code example that reproduces the error?

**Comment by martin-georgiev** [comment]
``` <?php

Route::bind('issue', function ($id) {
    return Registry::getManagerForClass(\MG\Entities\Issue\GeneralIssue::class)->
            getRepository(\MG\Entities\Issue\GeneralIssue::class)->find($id);
});
Route::resource('issue', \MG\Http\Controllers\Issue\IssueController::class);
Route::get('issue/export', \MG\Http\Controllers\Issue\IssueController::class.'@export');
```

**Comment by maxbrokman** [comment]
What does your RouteServiceProvider look like? We can reproduce your problem with and without LD by combining a route group with a namespace and `Controller::class`.

**Comment by martin-georgiev** [comment]
I suspect that the root of the problem is in Laravel itself, but still something in LD activates it.  

```
class RouteServiceProvider extends ServiceProvider
{
    /**
     * @var string
     */
    protected $namespace = 'MG\Http\Controllers';

    public function map(Router $router)
    {
        $router->group(['namespace' => $this->namespace], function($router)
        {
            require app_path('Http/routes.php');
        });
    }
}
```

**Comment by FoxxMD** [comment]
@martin-georgiev Are you sure you are using `Route::resource` correctly? In the [documentation](http://laravel.com/docs/5.1/controllers#restful-resource-controllers) only the controller name is listed. I've debugged the router using a FQN as the controller name (as per your issue) and here's how it works:

At **ResourceRegistrar@register** line 60

``` php
        foreach ($this->getResourceMethods($defaults, $options) as $m) {
            $this->{'addResource'.ucfirst($m)}($name, $base, $controller, $options);
        }
```

`$controller` is the FQN passed to the route (`App\Htttp\Controllers\UserController`)

At **ResourceRegistrar@addResourceIndex** (and this occurs for all other methods)

``` php
    protected function addResourceIndex($name, $base, $controller, $options)
    {
        $uri = $this->getResourceUri($name);

        $action = $this->getResourceAction($name, $controller, 'index', $options);

        return $this->router->get($uri, $action);
    }
```

**ResourceRegistrar@getResourceAction** line 175

``` php
        return ['as' => $name, 'uses' => $controller.'@'.$method];
```

It returns an action with `uses` set to `App\Htttp\Controllers\UserController@index`

The `return` for `addResourceIndex` eventually leads us to **Router@createRoute** line 488

``` php
        if ($this->actionReferencesController($action)) {
            $action = $this->convertToControllerAction($action);
        }
```

`actionReferencesController` checks to see if `action` is an array and has `uses` set (which it does, from `getResourceAction`) so it calls `convertToControllerAction`

``` php
protected function convertToControllerAction($action)
    {
        if (is_string($action)) {
            $action = ['uses' => $action];
        }

        // Here we'll merge any group "uses" statement if necessary so that the action
        // has the proper clause for this property. Then we can simply set the name
        // of the controller on the action and return the action array for usage.
        if (! empty($this->groupStack)) {
            $action['uses'] = $this->prependGroupUses($action['uses']);
        }

        // Here we will set this controller name on the action array just so we always
        // have a copy of it for reference if we need it. This can be used while we
        // search for a controller name or do some other type of fetch operation.
        $action['controller'] = $action['uses'];

        return $action;
    }
```

since `$action` is not a string it does not use the controller as is. It moves to `groupStack` which is not empty and calls `prependGroupUses`

``` php
    protected function prependGroupUses($uses)
    {
        $group = end($this->groupStack);

        return isset($group['namespace']) && strpos($uses, '\\') !== 0 ? $group['namespace'].'\\'.$uses : $uses;
    }
```

This is where the namespace gets prepended. If you have set `namespace` in `RouteServiceProvider` then there is always a group and that namespace is always prepended. 

@maxbrokman , myself, and @patrickbrouwers reproduced your problem but it was not corrected by reverting to 1.0.6.

@patrickbrouwers was also able to reproduce without using LD by doing a clean install of Laravel 5.1.22 with the route

``` php
use App\Http\Controllers\HomeController;

Route::get('/', HomeController::class . '@index’);
```

I'm not sure how you were able to use a FQN as a controller name before we're pretty dang sure it wasn't laravel doctrine :) Hope this helps

**Comment by patrickbrouwers** [comment]
Will close this, as I don't it's related to Doctrine.
